### PR TITLE
Add nn.py to the repository

### PR DESCRIPTION
**Pull request by NicolasLegendre1**

### CODE REVIEW

File: reduceSPI/nn.py [hunk 1]
```diff
@@ -0,0 +1,890 @@
...
+    3: nn.Conv3d}
+NN_CONV_TRANSPOSE = {
+    2: nn.ConvTranspose2d,
+    3: nn.ConvTranspose3d}
+
+NN_BATCH_NORM = {
+    2: nn.BatchNorm2d,
+    3: nn.BatchNorm3d}
+
+LATENT_SPACE_DEFINITION = {
```
Comment (ninamiolane) [inline reduceSPI/nn.py:54]: This should be a list, it would automatically index the elements with the indices 0-4

File: reduceSPI/nn.py [hunk 2]
```diff
@@ -0,0 +1,890 @@
...
+    CONSTANTS["latent_space_definition"] = 0
+    CONSTANTS["latent_dim"] = 10
+    CONSTANTS["skip_z"] = True
+    CONSTANTS["n_gan_lay"] = 1
+    enc = EncoderConv(CONSTANTS)
+    A = torch.zeros(20, 1, 64, 64)
+    B = enc.forward(A)
+    dec = DecoderConv(CONSTANTS)
+    C = dec.forward(B[2])
+    dis = Discriminator(CONSTANTS)
```
Comment (ninamiolane) [inline reduceSPI/nn.py:36]: Are you using the GAN part for your experiments ? Let's try to do without it -- to simplify and get experiments running first.

File: reduceSPI/nn.py [hunk 3]
```diff
@@ -0,0 +1,890 @@
...
+    2: nn.ConvTranspose2d,
+    3: nn.ConvTranspose3d}
+
+NN_BATCH_NORM = {
+    2: nn.BatchNorm2d,
+    3: nn.BatchNorm3d}
+
+LATENT_SPACE_DEFINITION = {
+    0: "(R^L)",
+    1: "(SO(3)xT)",
```
Comment (ninamiolane) [inline reduceSPI/nn.py:56]: When you write SO(3), do you mean:
- the space of rotation matrices? 
- the Euler angles? (if so which one)
- the axis-angle reparameterization?

Comment (ninamiolane) [reply reduceSPI/nn.py:56]: Great, it would be nice to have cases without the translation too. (we want the translation in the end, but right now our priority is to get something that runs).

File: reduceSPI/nn.py [hunk 4]
```diff
@@ -0,0 +1,890 @@
...
+    2: "(SO(3)xT)z0->R^L",
```
Comment (ninamiolane) [inline reduceSPI/nn.py:57]: -> R^3 ?

File: reduceSPI/nn.py [hunk 5]
```diff
@@ -0,0 +1,890 @@
...
+
+        self.last_out_shape = next_in_shape
+
+        self.fcs_infeatures = functools.reduce(
+            (lambda x, y: x * y), self.last_out_shape)
+
+        self.fc_mu, self.fc_logvar = self.init_layer_latent_space()
+
+    def init_layer_latent_space(self):
+        if self.latent_space_definition == 0:
```
Comment (ninamiolane) [inline reduceSPI/nn.py:248]: Directly test with the equality of strings, it would help reading the code without having to go back to the beginning of the file to remember which index represents which case 🤗

File: reduceSPI/nn.py [hunk 6]
```diff
@@ -0,0 +1,890 @@
...
+        elif self.latent_space_definition in [1, 2, 3]:
+            mu = nn.Linear(
+                in_features=self.fcs_infeatures, out_features=5)
+            logvar = nn.Linear(
+                in_features=self.fcs_infeatures, out_features=5)
+        elif self.latent_space_definition == 4:
+            mu = nn.Linear(
+                in_features=self.fcs_infeatures, out_features=8)
+            logvar = nn.Linear(
+                in_features=self.fcs_infeatures, out_features=5)
```
Comment (ninamiolane) [inline reduceSPI/nn.py:262]: remind me why we have 5 and not 8 here?

File: reduceSPI/nn.py [hunk 7]
```diff
@@ -0,0 +1,890 @@
...
+        wig_matrix = convert_rot_mat_ten_to_wig_mat_ten(
+            rot_mats, self.latent_dim)
+        z0 = torch.matmul(wig_matrix, self.z0)
+        z1 = torch.matmul(wig_matrix, self.z1)
+        z = torch.cat((z0, z1, translation), dim=1)
+        return mu, logvar, z, rot_mats
+
+
+def reparametrize_so3(mu, logvar):
+    mu_rotmat = mu.T[:-2].T
```
Comment (ninamiolane) [inline reduceSPI/nn.py:330]: calling this rotmat is confusing because this is not a matrix. (I know the code is not clean, but it can help me read faster through it if the var names are intuitive).

mu_rot ?

File: reduceSPI/nn.py [hunk 8]
```diff
@@ -0,0 +1,890 @@
...
+        self.last_out_shape = next_in_shape
+
+        self.fcs_infeatures = functools.reduce(
+            (lambda x, y: x * y), self.last_out_shape)
+
+        self.fc_mu, self.fc_logvar = self.init_layer_latent_space()
+
+    def init_layer_latent_space(self):
+        if self.latent_space_definition == 0:
+            mu = nn.Linear(
```
Comment (ninamiolane) [inline reduceSPI/nn.py:249]: dangerous to call this mu, because we might mix up the mu and the layer generative the mu. How about fc_mu and fc_logvar?

File: reduceSPI/nn.py [hunk 9]
```diff
@@ -0,0 +1,890 @@
...
+        eps = torch.FloatTensor(n_samples, n_batch_data,
+                                latent_shape).normal_()
+    eps = eps.reshape(eps.shape[1:])
+    tang_mu = eps*sigma
+    z = mu+tang_mu
+    matrices = rot_mat_tensor(z, latent_dim)
+    return matrices
+
+
+def rot_mat_tensor(tens, latent_dim):
```
Comment (ninamiolane) [inline reduceSPI/nn.py:447]: from_euler_angles_to_rotmat ?

It does not seem that this is generating a rotation matrix either - can you explain a bit what is happening here?

File: reduceSPI/nn.py [hunk 10]
```diff
@@ -0,0 +1,890 @@
...
+    batch_size = tens.shape[0]
+    J = Jd[int((latent_dim-1)/2)].to(DEVICE)
+    matrices = torch.Tensor(batch_size, latent_dim, latent_dim).to(DEVICE)
+    for i in range(batch_size):
+        alpha = tens[i][0]
+        beta = tens[i][1]
+        gamma = tens[i][2]
+        matrices[i] = rot_mat(alpha, beta, gamma, int((latent_dim-1)/2), J)
```
Comment (ninamiolane) [inline reduceSPI/nn.py:455]: what does the rot_mat function (imported from another module) do?

File: reduceSPI/nn.py [hunk 11]
```diff
@@ -0,0 +1,890 @@
...
+        z = torch.cat((z0, z1, translation), dim=1)
+        return mu, logvar, z, rot_mats
+
+
+def reparametrize_so3(mu, logvar):
+    mu_rotmat = mu.T[:-2].T
+    mu_transl = mu.T[-2:].T
+    logvar_rotmat = logvar.T[:-2].T
+    logvar_transl = logvar.T[-2:].T
+    rot_mats_no_var = rot_mat_tensor(mu_rotmat, 3)
```
Comment (ninamiolane) [inline reduceSPI/nn.py:334]: Sorry I'm having a hard time to follow here

File: reduceSPI/nn.py [hunk 12]
```diff
@@ -0,0 +1,890 @@
...
+    mu_rotmat = mu.T[:-2].T
+    mu_transl = mu.T[-2:].T
+    logvar_rotmat = logvar.T[:-2].T
+    logvar_transl = logvar.T[-2:].T
+    translation = reparametrize(mu_transl, logvar_transl)
+    rot_mats_no_var = rotation_vectors(mu_rotmat)
+    return rot_mats_no_var, translation, logvar_rotmat
+
+
+def rotation_vectors(z):
```
Comment (ninamiolane) [inline reduceSPI/nn.py:349]: `def from_s2s2_to_rotmat`

File: reduceSPI/nn.py [hunk 13]
```diff
@@ -0,0 +1,890 @@
...
+    z = z.reshape((-1, 6))
+    u = z.T[:3].T
+    v = z.T[3:].T
+    e1 = nn.functional.normalize(u)
+    v2 = v - (e1*v).sum(-1, keepdim=True) * e1
+    e2 = nn.functional.normalize(v2)
+    e3 = torch.cross(e1, e2)
+    matrices = torch.cat([e1, e2, e3]). reshape((-1, 3, 3))
```
Comment (ninamiolane) [inline reduceSPI/nn.py:357]: are we sure this is happening correctly, i.e. we get the matrix R that we want and not, say, R^T?

A unit test for this function would be good.

File: reduceSPI/nn.py [hunk 14]
```diff
@@ -0,0 +1,890 @@
...
+    return matrices
+
+
+def map_to_lie_algebra(v):
```
Comment (ninamiolane) [inline reduceSPI/nn.py:361]: This function needs a unit-test, to make sure that we do not make a mistake here (I'm not a good unit-tester just by reading)

Comment (ninamiolane) [reply reduceSPI/nn.py:361]: def from_vector_to_skewsym_mat

File: reduceSPI/nn.py [hunk 15]
```diff
@@ -0,0 +1,890 @@
...
+                        [1., 0., 0.],
+                        [0., 0., 0.]])
+
+    R = R_x * v[..., 0, None, None] + \
+        R_y * v[..., 1, None, None] + \
+        R_z * v[..., 2, None, None]
+    return R
+
+
+def rodrigues(v):
```
Comment (ninamiolane) [inline reduceSPI/nn.py:389]: A unit test for this too

File: reduceSPI/nn.py [hunk 16]
```diff
@@ -0,0 +1,890 @@
...
+    # normalize K
+    K = map_to_lie_algebra(v / theta)
+
+    I = torch.eye(3, device=v.device, dtype=v.dtype)
+    R = I + torch.sin(theta)[..., None]*K \
+        + (1. - torch.cos(theta))[..., None]*(K@K)
+    return R
+
+
+def add_logvar(rot_mats_no_var, logvar_rotmat):
```
Comment (ninamiolane) [inline reduceSPI/nn.py:400]: There is no sampling here, and there should be.

1) use logvar to sample a vector in R^3
2) and then rodrigues of that vector

right now you are using the value of the var to represent the vector that should be sampled.

File: reduceSPI/nn.py [hunk 17]
```diff
@@ -0,0 +1,890 @@
...
+    logvar_mat = rodrigues(logvar_rotmat)
+    return rot_mats_no_var @ logvar_mat
+
+
+def convert_rot_mat_to_eul_ang(matrix):
+    y1, y2, z = 0, 0, 0
```
Comment (ninamiolane) [inline reduceSPI/nn.py:406]: i dont think you need this init?

File: reduceSPI/nn.py [hunk 18]
```diff
@@ -0,0 +1,890 @@
...
+        + (1. - torch.cos(theta))[..., None]*(K@K)
+    return R
+
+
+def add_logvar(rot_mats_no_var, logvar_rotmat):
+    logvar_mat = rodrigues(logvar_rotmat)
+    return rot_mats_no_var @ logvar_mat
+
+
+def convert_rot_mat_to_eul_ang(matrix):
```
Comment (ninamiolane) [inline reduceSPI/nn.py:405]: which euler representation is used?

Comment (ninamiolane) [reply reduceSPI/nn.py:405]: we need a unit test for this

File: reduceSPI/nn.py [hunk 19]
```diff
@@ -0,0 +1,890 @@
...
+        y2 = math.atan2(matrix[1][2], matrix[1][0])
+        z = math.atan2(sy, matrix[1][1])
+    else:
+        y1 = 0
+        y2 = math.atan2(matrix[1][0], matrix[0][0])
+        z = 0
+    return torch.Tensor([y1, z, y2])
+
+
+def convert_rot_mat_ten_to_wig_mat_ten(matrices, latent_dim):
```
Comment (ninamiolane) [inline reduceSPI/nn.py:419]: need a unit test here

File: reduceSPI/nn.py [hunk 20]
```diff
@@ -0,0 +1,890 @@
...
+    for i in range(batch_size):
+        euler_tens[i] = convert_rot_mat_to_eul_ang(matrices[i])
+    wig_matrices = rot_mat_tensor(euler_tens, latent_dim)
+    return wig_matrices
+
+
+SO3 = SpecialOrthogonal(3, point_type="vector")
+
+
+def transform_into_so3(mu, logvar, latent_dim, n_samples=1):
```
Comment (ninamiolane) [inline reduceSPI/nn.py:431]: This is not used? maybe remove for this PR?

**Review by ninamiolane — COMMENTED** [review]
Nice, thanks for sharing! I made a few comments

File: reduceSPI/nn.py (hunk 6 above)
Comment (NicolasLegendre1) [reply reduceSPI/nn.py:262]: because to compute the variation matrix you only need 3 coordonnates. So 5 = 3+2 (rotation+ translation)

File: reduceSPI/nn.py (hunk 9 above)
Comment (NicolasLegendre1) [reply reduceSPI/nn.py:447]: this function is generation a wigner matrix
 If you choose latent_dim = 3 you obtain exactly the rotation matrix

File: reduceSPI/nn.py (hunk 10 above)
Comment (NicolasLegendre1) [reply reduceSPI/nn.py:455]: this function compute the wigner matrix

File: reduceSPI/nn.py (hunk 11 above)
Comment (NicolasLegendre1) [reply reduceSPI/nn.py:334]: I change the name of variables it will probably be easier to understand

File: reduceSPI/nn.py (hunk 13 above)
Comment (NicolasLegendre1) [reply reduceSPI/nn.py:357]: yes the idea now is to obtain a matrix of rotation, I need after that to convert it into wigner matrix and then to multiply the matrix with a vector of R^T

File: reduceSPI/nn.py (hunk 15 above)
Comment (NicolasLegendre1) [reply reduceSPI/nn.py:389]: I took this function from the git of the paper you sent me
